### DISCUSSION TITLE
Differences between previewer `bat` and `bat_native`?

### DISCUSSION THREAD

**Discussion by josh-nz** [ibhagwan/fzf-lua#1110] [Q&A] [answered]
I'm wondering what the differences are between these two previewer settings. The wiki page reads:

```
the difference between the two is that the former [bat] will use neovim as a proxy to the bat command and bat_native will defer the execution of bat to fzf (hence "native").
```

I would assume that `bat_native` is the fastest, but if this is correct, why then do the `fzf-native` and `max-perf` profiles use `bat` and not `bat_native` (but do use `{man|help}_native`)?

Aside from speed, are there any other things to consider when deciding between these two settings?

**Comment by ibhagwan** [comment, marked as answer]
The “native” as its name suggests sets fzf’s preview command directly to `bat <args> <filename>`, however for filename is then replaced with a `neovim --headless --cmd <lua>` command in order to properly extract the filename from the entry (strip icons, etc) - this approach has obvious limitations with tags for example where the entry doesn’t have a line number and other issues with Windows where the files index for the line number differs between relative paths and absolute paths (due to the extra `C:` prefix when our delimiter is `:`).

The non native version uses a `neovim --headless …` directly as fzf’s `--preview` argument and then spawns bat using libuv’s spawn and streams the data over named pipe to fzf.

Both technically are a separate process with the non-native doing a bit more work on neovim’s main thread as opposed to the native version but the perf difference is probably not that big.

> why then do the fzf-native and max-perf profiles use bat and not bat_native (but do use {man|help}_native

`bat` is technically already counted as a “native” previewer, meaning the preview is spawned in the terminal inside fzf’s preview “window”, the reason it’s the default is due to the limitations I explained above (namely `tags` / Windows) and the fact I’d like to remove it one day :)

**Comment by ibhagwan** [comment]
Btw, if you want to see this in action, use `:FzfLua files previewer=bat debug=true` and `:FzfLua files previewer=bat_native debug=true` and then explore the generated fzf cli in `:messages` and inspect the `--preview` arg
> bat_native
```sh

--preview 'bat --color=always --style=numbers,changes  "$(VIMRUNTIME='\''/usr/share/nvim/ru
ntime'\'' '\''/usr/bin/nvim'\'' -n --headless --clean --cmd '\''lua vim.g.did_load_filetypes=1; loadfile([[/home/bhagwan/Sources/nvim/fzf-lua/lua/fzf-lua/shell_helper.lu
a]])().rpc_nvim_exec_lua({fzf_lua_server=[[/run/user/1000/fzf-lua.1712038571.10020.1]], fnc_id=5 , debug=true})'\'' -- {})"'
```
> bat
```sh

--preview 'VIMRUNTIME='\''/usr/share/nvim/runtime'\'' '\''/usr/bin/nvim'\'' -n --headless -
-clean --cmd '\''lua vim.g.did_load_filetypes=1; loadfile([[/home/bhagwan/Sources/nvim/fzf-lua/lua/fzf-lua/shell_helper.lua]])().rpc_nvim_exec_lua({fzf_lua_server=[[/run
/user/1000/fzf-lua.1712038571.10020.1]], fnc_id=6 , debug=true})'\'' -- {})"'
```
